### ISSUE TITLE
Game Crashes with TerraFirmaCraft and DragonAPI V25c

### ISSUE THREAD

**Issue by Superfrogman98** [ReikaKalseki/Reika_Mods_Issues#2396]
Upon updating the Reika mods in my modpack, I get a crash on startup. I made a smaller pack to reduce the possible interactions causing problems.   

Mods:
-1.7.10-TerraFirmaCraft-0.79.30.925
-DragonAPI 1.7.10 V25c

Crash Report Is as Follows: https://gist.github.com/Superfrogman98/b11abd4927f203fe28cf434b78ab0d31

**Comment by ReikaKalseki** [comment]
Who develops TFC? Does it ASM this same class and method?

`Caused by: Reika.DragonAPI.Exception.ASMException$NoSuchASMMethodInstructionException: Error ASMing method func_78464_a ()V in net/minecraft/client/renderer/EntityRenderer:`

**Comment by Superfrogman98** [comment]
dries007, Bunsan, and AlcatrazEscapee are the current Devs. 
From the looks of their code, It appears that they do use ASM to change that class so it uses their own method in place of vanillas.

**Comment by ReikaKalseki** [comment]
I do not see a good fix for this, especially given they are not likely updating the code for 1.7; I could easily make it override their patch, but that will break something on their end, and just making it skip the patch breaks things on mine. And I cannot dynamically call their code, either, as reflection in the render pipeline is going to negatively impact FPS.

**Comment by Superfrogman98** [comment]
I looked at the specific changes they do to that class a bit more, and it looks like they only modify updateRenderer so that it calls their method doRainClient in place of the default addRainParticles. Looking at the differences from the vanilla function I believe this is so they can use their weather manager and climates for where rain/snow splash particles should be shown. This would probably cause some odd splash rendering mixups during rain/snow as seasons change in the more temperate zones, but compared to the fixes and additions in your updates, this seems like a very minor problem.

This is what they are calling in place of the vanilla method  (lines 74-160)
https://github.com/TerraFirmaCraft/TFC-1.7.10/blob/master/src/ASM/com/bioxx/tfc/ClientOverrides.java

**Comment by ReikaKalseki** [comment]
Right, but I cannot make my patched code call their method without reflection.

**Comment by Superfrogman98** [comment]
I was able to set up a dev environment for TFC, so I could remove the ASM for rain so it doesn't crash. I then used your NORAINFX config to disable the default particles/sound, and added an optional config to my custom content mod for the pack for enabling an event based way of making the particles/sound work as they should with TFC. Rather then directly distributing this modified version of TFC, I intend to use a mod created by the Gregtech: New Horizons team to apply the patch after the mod is installed from the official source. With your rules for modpack distrubution being as they are, I was wondering if this type of fix is permitted, as long as it isn't your mods that I patch?

**Comment by ReikaKalseki** [comment]
My rules only apply to my mods. That said, ensure that TFC/Greg are OK with this.

**Comment by OvermindDL1** [comment]
> Right, but I cannot make my patched code call their method without reflection.

Your patched code can call whatever it wants, you can asm the call straight in to your code so there is no virtual/reflection overhead.  More easily would be calling a jump method on a class that you instance once globally and just no-op if it is unable to be loaded due to TFC's call not existing.  That's what GT6 does for most of such calls and the JIT optimizes it very well so not even any virtual overhead.

**Comment by ReikaKalseki** [comment]
I am not willing to build the entire method with ASM due to the complexity involved. I only am willing to use ASM to redirect methods.

**Comment by OvermindDL1** [comment]
Wouldn't need to build the whole method with asm, just build it as normal but call a no-op shim, and just replace your shim call with the real call via ASM, which is a simple swap of the name in the opcode.  ^.^
